test/Cut.hs

```diff
@@ -22,34 +26,40 @@ tests = testGroup "Cut"
     [ testMonad
     -- , testMonadFix
     , testCut
-    ] >>= ($ RunL CutC.runCutA)
+    ] >>= ($ runL CutC.runCutA)
+  , testGroup "ReaderC · CutC" $
+    Cut.test (local (id @R)) (m (gen0 S.<> Reader.gen0 r) (\ m -> genN m S.<> Reader.genN r m)) a b (pair <*> r <*> unit) (Run (CutC.runCutA . uncurry runReader))
+  -- , testGroup "CutC · ReaderC" $
+  --   Cut.test (local (id @R)) (m (gen0 S.<> Reader.gen0 r) (\ m -> genN m S.<> Reader.genN r m)) a b (pair <*> r <*> unit) (Run (uncurry ((. CutC.runCutA) . runReader)))
   ] where
-  testMonad    run = Monad.test    (m gen) a b c (identity <*> unit) run
-  -- testMonadFix run = MonadFix.test (m gen) a b   (identity <*> unit) run
-  testCut      run = Cut.test      (m gen) a b                       run
+  testMonad    run = Monad.test    (m gen0 genN) a b c initial run
+  -- testMonadFix run = MonadFix.test (m gen0 genN) a b   initial run
+  testCut      run = Cut.test id   (m gen0 genN) a b   initial run
+  initial = identity <*> unit
 
 
-gen :: (Has Cut sig m, Has NonDet sig m) => GenM m -> GenM m
-gen m a = choice
-  [ label "call" call <*> m a
-  , label "cutfail" cutfail
-  , NonDet.gen m a
-  ]
+gen0 :: (Has Cut sig m, Has NonDet sig m) => Gen a -> [Gen (m a)]
+gen0 a = label "cutfail" cutfail : NonDet.gen0 a
+
+genN :: (Has Cut sig m, Has NonDet sig m) => GenM m -> Gen a -> [Gen (m a)]
+genN m a = (label "call" call <*> m a) : NonDet.genN m a
 
 
 test
-  :: forall a b m sig
-  .  (Has Cut sig m, Has NonDet sig m, Arg a, Eq a, Eq b, Show a, Show b, Vary a)
-  => GenM m
+  :: forall a b m f sig
+  .  (Has Cut sig m, Has NonDet sig m, Arg a, Eq a, Eq b, Show a, Show b, Vary a, Functor f)
+  => (forall a . m a -> m a)
+  -> GenM m
   -> Gen a
   -> Gen b
-  -> RunL [] m
+  -> Gen (f ())
+  -> Run f [] m
   -> [TestTree]
-test m a b (RunL runCut)
-  = testProperty "cutfail annihilates >>=" (forall (fn @a (m a) :. Nil)
-    (\ k -> runCut (cutfail >>= k) === runCut cutfail))
-  : testProperty "cutfail annihilates <|>" (forall (m a :. Nil)
-    (\ m -> runCut (cutfail <|> m) === runCut cutfail))
-  : testProperty "call delimits cutfail" (forall (m a :. Nil)
-    (\ m -> runCut (call cutfail <|> m) === runCut m))
-  : NonDet.test m a b (RunL runCut)
+test hom m a b i (Run runCut)
+  = testProperty "cutfail annihilates >>=" (forall (i :. fn @a (m a) :. Nil)
+    (\ i k -> runCut ((hom cutfail >>= k) <$ i) === runCut (hom cutfail <$ i)))
+  : testProperty "cutfail annihilates <|>" (forall (i :. m a :. Nil)
+    (\ i m -> runCut ((hom cutfail <|> m) <$ i) === runCut (hom cutfail <$ i)))
+  : testProperty "call delimits cutfail" (forall (i :. m a :. Nil)
+    (\ i m -> runCut ((hom (call (hom cutfail)) <|> m) <$ i) === runCut (m <$ i)))
+  : NonDet.test m a b i (Run runCut)
```

Comment:
As described above, homomorphisms are applied to each action. One alternative I’m considering (but haven’t explored yet) is to define a polymorphic carrier which `hmap`s a function over all of the positions in an effect before delegating the effect to its underlying carrier.